packages/go_router/lib/src/route.dart

```diff
@@ -446,22 +473,201 @@ class ShellRoute extends RouteBase {
+///         StackedNavigationItem(
+///           rootRoutePath: '/b', navigatorKey: _tabBNavigatorKey),
+///       ],
+///       scaffoldBuilder: (BuildContext context, int currentIndex,
+///           List<StackedNavigationItemState> itemsState,
+///           Widget scaffoldBody) {
+///         return ScaffoldWithBottomNavigationBar(
+///             currentIndex: currentIndex,
+///             itemsState: itemsState,
+///             body: scaffoldBody,
+///         );
+///       },
+///       routes: <RouteBase>[
+///         /// The root of tab 'A'
+///         GoRoute(
+///           path: '/a',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'A', detailsPath: '/a/details'),
+///         ),
+///         /// The root of tab 'B'
+///         GoRoute(
+///           path: '/b',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'B', detailsPath: '/b/details/1'),
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_nested_navigation.dart)
+/// for a complete runnable example.
+class PartitionedShellRoute extends ShellRouteBase {
+  /// Constructs a [PartitionedShellRoute] from a list of routes, each
+  /// representing the root of a navigation hierarchy. A separate
+  /// [Navigator] will be created for each of the root routes, using the
+  /// navigator key at the corresponding index in [navigatorKeys].
+  PartitionedShellRoute({
+    required List<RouteBase> routes,
+    required this.navigatorKeys,
+    super.builder,
+    super.pageBuilder,
+  })  : assert(routes.isNotEmpty),
+        assert(navigatorKeys.length == routes.length),
+        assert(Set<GlobalKey<NavigatorState>>.from(navigatorKeys).length ==
+            routes.length),
+        assert(pageBuilder != null || builder != null,
+            'builder or pageBuilder must be provided'),
+        super._(routes: routes) {
+    for (int i = 0; i < routes.length; ++i) {
+      final RouteBase route = routes[i];
+      if (route is GoRoute) {
+        assert(route.parentNavigatorKey == null ||
+            route.parentNavigatorKey == navigatorKeys[i]);
+      }
+    }
+  }
+
+  /// Constructs a [PartitionedShellRoute] that places its navigators in an
+  /// [IndexedStack], managed by a [StackedNavigationShell].
+  ///
+  /// Each route in the `routes` parameter must correspond to a
+  /// [StackedNavigationItem], specified in the `stackItems` parameter.
+  /// The stacked navigation shell can be implemented by specifying a
+  /// `scaffoldBuilder`, to build a widget that wraps the indexed stack.
+  factory PartitionedShellRoute.stackedNavigationShell({
+    required List<RouteBase> routes,
+    required List<StackedNavigationItem> stackItems,
+    StackedNavigationScaffoldBuilder? scaffoldBuilder,
```

Comment:
The Stacked keyword is a bit confusing. It only make sense if I understand it uses Stack to preservestate. However, the usage of stack is just an implementation detail. There can be different way to preserveState, and the users probably don't care as long as the state is preserved.

I suggest we remove all the mention of stack or index stack in the documentation and variable naming, and replace them with something along the line of `preserve state` or `stateful` or `persistent` or `maintain state`

Reply:
👍